File: tests/e2e/job_iterations_test.go
package e2e

import (
	"context"
	"sync/atomic"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/icegatetech/jobmanager"
	"github.com/icegatetech/jobmanager/tests/e2e/testenv"
)

// TestJobIterations verifies that a job can complete and restart for multiple iterations.
func TestJobIterations(t *testing.T) {
	ctx := context.Background()

	// 1. Start MinIO
	minioEnv, err := testenv.NewMinIOEnv(ctx, t)
	require.NoError(t, err)
	defer func() {
		if err := minioEnv.Close(); err != nil {
			t.Logf("failed to close MinIO: %v", err)
		}
	}()

	// 2. Track iterations
	expectedIterations := uint64(3)
	var iterationCount atomic.Uint64

	taskDef, err := jobmanager.NewTaskDefinition("iteration_task", nil)
	require.NoError(t, err)

	executor := func(ctx context.Context, task jobmanager.ImmutableTask, jm jobmanager.JobManager) error {
		current := iterationCount.Add(1)
		t.Logf("Executing iteration %d", current)

		// Complete the task - job will automatically restart for next iteration
		return jm.CompleteTask(task.ID(), []byte("done"))
	}

	jobDef, err := jobmanager.NewJobDefinition(
		"test_iterations_job",
		[]jobmanager.TaskDefinition{taskDef},
		map[jobmanager.TaskCode]jobmanager.TaskExecutor{
			"iteration_task": executor,
		},
		jobmanager.WithMaxIterations(expectedIterations),
	)
	require.NoError(t, err)

	jobDefs, err := jobmanager.NewJobDefinitions(jobDef)
	require.NoError(t, err)

	// 3. Create storage
	logger := testenv.NewTestLogger(t)
	storage, err := jobmanager.NewS3Storage(
		ctx,
		jobmanager.S3StorageConfig{
			Endpoint:        minioEnv.Endpoint(),
			AccessKeyID:     minioEnv.Username(),
			SecretAccessKey: minioEnv.Password(),
			BucketName:      "test-jobs",
			UseSSL:          false,
			BucketPrefix:    "jobs/",
		},
		logger,
		jobmanager.NewDisabledMetrics(),
		jobDefs,
		jobmanager.NewRetrier(jobmanager.RetrierConfig{}, logger),
	)
	require.NoError(t, err)

	// 4. Start manager
	managerEnv, err := testenv.NewManagerEnv(
		t,
		storage,
		logger,
		jobDefs,
		jobmanager.JobsManagerConfig{
			WorkerCount: 1,
		},
	)
	require.NoError(t, err)

	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	err = managerEnv.Start(ctx)
	require.NoError(t, err)

	// 5. Wait for all iterations to complete
	err = managerEnv.WaitForAllJobsCompletion(ctx, 15*time.Second)
	require.NoError(t, err)
	cancel()
	managerEnv.Wait()

	// 6. Verify correct number of iterations
	assert.Equal(t, expectedIterations, iterationCount.Load(), "should have completed all iterations")

	// Verify final job state
	job, err := storage.GetJob(context.Background(), "test_iterations_job")
	require.NoError(t, err)
	assert.Equal(t, jobmanager.JobCompleted, job.Status())
	assert.Equal(t, expectedIterations, job.IterationNum(), "job should be at final iteration")
}

File: src/tests/job_iterations_test.rs
use std::{
    sync::{
        Arc,
        atomic::{AtomicI64, AtomicU64, Ordering},
    },
    time::Duration,
};

use chrono::{Duration as ChronoDuration, Utc};
use tokio_util::sync::CancellationToken;

use super::common::manager_env::ManagerEnv;
use super::common::s3_container::S3TestContainer;
use crate::storage::in_memory::InMemoryStorage;
use crate::{
    JobCode, JobDefinition, JobDefinitionId, JobRegistry, JobStateCodecKind, JobStatus, JobsManagerConfig, NoopMetrics,
    S3Storage, S3StorageConfig, TaskCode, TaskDefinition, TaskLimits, TaskOutcome, task_fn,
};

/// `TestJobIterations` verifies that a job can complete and restart for multiple iterations
#[tokio::test]
async fn test_job_iterations() -> Result<(), Box<dyn std::error::Error>> {
    super::common::init_tracing();

    // 1. Start object storage
    let store = S3TestContainer::start().await?;

    // 2. Track iterations
    let expected_iterations = 3u64;
    let iteration_count = Arc::new(AtomicU64::new(0));

    let iteration_count_clone = Arc::clone(&iteration_count);

    let executor = task_fn(move |_ctx| {
        let count = Arc::clone(&iteration_count_clone);

        async move {
            let current = count.fetch_add(1, Ordering::SeqCst) + 1;
            tracing::info!("Executing iteration {}", current);

            // Complete the task - job will automatically restart for next iteration
            Ok(TaskOutcome::Completed(b"done".to_vec()))
        }
    });

    let task_def = TaskDefinition::new(TaskCode::new("iteration_task"), Duration::from_secs(5));

    let job_def = JobDefinition::new(
        JobDefinitionId::new(),
        JobCode::new("test_iterations_job"),
        vec![(task_def, executor)],
        Vec::new(),
        Vec::new(),
        TaskLimits::default(),
    )?
    .with_max_iterations(expected_iterations)?;

    // 3. Create job definitions
    let job_registry = Arc::new(JobRegistry::new(vec![job_def.clone()])?);

    // 4. Create storage
    let storage = S3Storage::new(
        S3StorageConfig::new(
            store.endpoint(),
            store.username(),
            store.password(),
            "test-jobs",
            "us-east-1",
        )
        .with_job_state_codec(JobStateCodecKind::Json),
        job_registry.clone(),
        Arc::new(NoopMetrics),
    )
    .await?;

    // 5. Start manager
    let config = JobsManagerConfig {
        worker_count: 1,
        worker_config: super::common::build_worker_config(Duration::from_millis(100), Duration::from_millis(10)),
        ..Default::default()
    };

    let mut manager_env = ManagerEnv::new(Arc::new(storage), config, Arc::clone(&job_registry), vec![job_def])?;

    // 6. Wait for all iterations to complete
    manager_env.wait_for_all_jobs_completion(Duration::from_secs(15)).await?;
    manager_env.stop().await;

    // 7. Verify correct number of iterations
    assert_eq!(
        iteration_count.load(Ordering::SeqCst),
        expected_iterations,
        "should have completed all iterations"
    );

    // Verify final job state
    let cancel_token = CancellationToken::new();
    let job = manager_env
        .storage()
        .get_job(&JobCode::new("test_iterations_job"), &cancel_token)
        .await?;
    assert_eq!(*job.status(), JobStatus::Completed);
    assert_eq!(job.iter_num(), expected_iterations, "job should be at final iteration");

    Ok(())
}

#[tokio::test]
async fn test_job_iterations_honors_next_start_at() -> Result<(), Box<dyn std::error::Error>> {
    super::common::init_tracing();

    let expected_iterations = 2u64;
    let delay_ms = 600i64;
    let iteration_count = Arc::new(AtomicU64::new(0));
    let first_iteration_completed_at_ms = Arc::new(AtomicI64::new(0));
    let second_iteration_started_at_ms = Arc::new(AtomicI64::new(0));

    let iteration_count_clone = Arc::clone(&iteration_count);
    let first_at_clone = Arc::clone(&first_iteration_completed_at_ms);
    let second_at_clone = Arc::clone(&second_iteration_started_at_ms);
    let executor = task_fn(move |ctx| {
        let count = Arc::clone(&iteration_count_clone);
        let first_at = Arc::clone(&first_at_clone);
        let second_at = Arc::clone(&second_at_clone);

        async move {
            let current = count.fetch_add(1, Ordering::SeqCst) + 1;

            if current == 1 {
                first_at.store(Utc::now().timestamp_millis(), Ordering::SeqCst);
                ctx.job()
                    .set_next_start_at(Utc::now() + ChronoDuration::milliseconds(delay_ms))?;
            }
            if current == 2 {
                second_at.store(Utc::now().timestamp_millis(), Ordering::SeqCst);
            }

            Ok(TaskOutcome::Completed(b"done".to_vec()))
        }
    });

    let task_def = TaskDefinition::new(TaskCode::new("iteration_task"), Duration::from_secs(5));

    let job_def = JobDefinition::new(
        JobDefinitionId::new(),
        JobCode::new("test_next_start_at_job"),
        vec![(task_def, executor)],
        Vec::new(),
        Vec::new(),
        TaskLimits::default(),
    )?
    .with_max_iterations(expected_iterations)?;
    let job_registry = Arc::new(JobRegistry::new(vec![job_def.clone()])?);

    let storage = Arc::new(InMemoryStorage::new());
    let config = JobsManagerConfig {
        worker_count: 1,
        worker_config: super::common::build_worker_config(Duration::from_millis(20), Duration::ZERO),
        ..Default::default()
    };
    let mut manager_env = ManagerEnv::new(
        storage.clone() as Arc<dyn crate::Storage>,
        config,
        Arc::clone(&job_registry),
        vec![job_def],
    )?;

    tokio::time::timeout(Duration::from_secs(5), async {
        while iteration_count.load(Ordering::SeqCst) < 1 {
            tokio::time::sleep(Duration::from_millis(10)).await;
        }
    })
    .await?;

    tokio::time::sleep(Duration::from_millis(
        u64::try_from(delay_ms / 2).expect("delay_ms should be non-negative"),
    ))
    .await;
    assert_eq!(
        iteration_count.load(Ordering::SeqCst),
        1,
        "second iteration should not start before next_start_at"
    );

    manager_env.wait_for_all_jobs_completion(Duration::from_secs(10)).await?;
    manager_env.stop().await;

    assert_eq!(
        iteration_count.load(Ordering::SeqCst),
        expected_iterations,
        "job should complete exactly two iterations"
    );

    let first_at = first_iteration_completed_at_ms.load(Ordering::SeqCst);
    let second_at = second_iteration_started_at_ms.load(Ordering::SeqCst);
    assert!(first_at > 0, "first iteration timestamp should be captured");
    assert!(second_at > 0, "second iteration timestamp should be captured");
    assert!(
        second_at - first_at >= delay_ms,
        "second iteration should start no earlier than configured delay"
    );

    let cancel_token = CancellationToken::new();
    let job = manager_env
        .storage()
        .get_job(&JobCode::new("test_next_start_at_job"), &cancel_token)
        .await?;
    assert_eq!(*job.status(), JobStatus::Completed);
    assert_eq!(job.iter_num(), expected_iterations);

    Ok(())
}
